docs(transformer/styled-components): add comments about `CSSMinifier`

Dunqing opened the pull request:
Added some comments to describe `CSSMinifier`, and clarify why `quasis.set_len(expressions.len() + 1);` is safe.

Also, there is a `assert` in codegen to ensure that `quasis` is always one greater than `expressions`

https://github.com/oxc-project/oxc/blob/c003ebad60970d27aae48b488ff361ef2b612379/crates/oxc_codegen/src/gen.rs#L2091-L2092


In addition, I found that we can use `truncate()` instead of `set_len`. Both are the same, but `truncate` is a safe function.

overlookmotel reviewed (approved):
Apart from the typo Graphite's spotted (it *finally* got something right), all looks good. Thank you for documenting this.

overlookmotel commented on crates/oxc_transformer/src/plugins/styled_components.rs:
```diff
@@ -903,6 +940,59 @@ impl<'a> CssMinifier<'a> {
+    ///
+    /// ### Steps:
+    ///
+    /// 1. It initializes state variables to track whether it's inside a string,
+    ///   the current string character, and the depth of parentheses.
+    /// 2. It iterates through each byte of the CSS string:
+    ///   - If it encounters a string quote (`"` or `'`), it toggles the `in_string` state.
+    ///   - If it finds a placeholder (starting with `_` and followed by
+    ///     [CssMinifier::PLACEHOLDER_PREFIX]), it tries to parse the placeholder and
+    ///     keeps track of its index.
+    ///   - It skips comments (both block and line comments) and whitespace, compressing them
+    ///    as needed.
+    ///   - It handles escaped newlines by replacing them with a space.
+    /// 3. It collects the minified CSS parts into `new_raws` and tracks which
+    ///   expressions were kept in a `remaining_expression_indexes`.
```
The nittiest nit of all time!

```suggestion
    /// 1. It initializes state variables to track whether it's inside a string,
    ///    the current string character, and the depth of parentheses.
    /// 2. It iterates through each byte of the CSS string:
    ///   - If it encounters a string quote (`"` or `'`), it toggles the `in_string` state.
    ///   - If it finds a placeholder (starting with `_` and followed by
    ///     [CssMinifier::PLACEHOLDER_PREFIX]), it tries to parse the placeholder and
    ///     keeps track of its index.
    ///   - It skips comments (both block and line comments) and whitespace, compressing them
    ///     as needed.
    ///   - It handles escaped newlines by replacing them with a space.
    /// 3. It collects the minified CSS parts into `new_raws` and tracks which
    ///    expressions were kept in a `remaining_expression_indexes`.
```

Dunqing committed:
docs(transformer/styled-components): add comments about `CSSMinifier` (#12197)

Added some comments to describe `CSSMinifier`, and clarify why `quasis.set_len(expressions.len() + 1);` is safe.

Also, there is a `assert` in codegen to ensure that `quasis` is always one greater than `expressions`

https://github.com/oxc-project/oxc/blob/c003ebad60970d27aae48b488ff361ef2b612379/crates/oxc_codegen/src/gen.rs#L2091-L2092

In addition, I found that we can use `truncate()` instead of `set_len`. Both are the same, but `truncate` is a safe function.

overlookmotel commented on crates/oxc_transformer/src/plugins/styled_components.rs:
```diff
@@ -412,8 +412,15 @@ impl<'a> StyledComponents<'a, '_> {
         }
 
-        // SAFETY: `expressions` is always one less than `quasis`.
+        // SAFETY:
+        // This is safe because template literal has ensured that `quasis` always
+        // has one more element than `expressions`, and the `CSSMinifier` guarantees that
```
#12213
